src/solvers/scip.rs

```diff
@@ -0,0 +1,171 @@
+use russcip::model::ObjSense;
+use russcip::variable::VarType;
+
+use crate::variable::{UnsolvedProblem, VariableDefinition};
+use crate::{
+    constraint::ConstraintReference,
+    solvers::{ObjectiveDirection, ResolutionError, Solution, SolverModel},
+};
+use crate::{Constraint, Variable};
+
+/// The [SCIP](https://scipopt.org) solver,
+/// to be used with [UnsolvedProblem::using].
+pub fn scip(to_solve: UnsolvedProblem) -> SCIPProblem {
+    let mut problem = Model::new();
+
+    problem.include_default_plugins();
+    problem.create_prob("problem");
+    problem.hide_output();
+    problem.set_obj_sense(match to_solve.direction {
+        ObjectiveDirection::Maximisation => ObjSense::Maximize,
+        ObjectiveDirection::Minimisation => ObjSense::Minimize,
+    });
+
+    for (
+        var,
+        &VariableDefinition {
+            min,
+            max,
+            is_integer,
+            ref name,
+        },
+    ) in to_solve.variables.iter_variables_with_def()
+    {
+        let coeff = *to_solve
+            .objective
+            .linear
+            .coefficients
+            .get(&var)
+            .unwrap_or(&0.);
+        let var_type = match is_integer {
+            true => VarType::Integer,
+            false => VarType::Continuous,
+        };
+        problem.add_var(min, max, coeff, name.as_str().clone(), var_type);
+    }
+
+    SCIPProblem { problem }
+}
+
+/// A SCIP Model
+pub struct SCIPProblem {
+    problem: Model,
+}
+
+impl SolverModel for SCIPProblem {
+    type Solution = SCIPSolution;
+    type Error = ResolutionError;
+
+    fn solve(self) -> Result<Self::Solution, Self::Error> {
+        let vars = self.problem.get_vars();
+        self.problem.solve();
+        let sol = self.problem.get_best_sol();
+        let values = vars
+            .iter()
+            .map(|var| (var.get_index(), sol.get_var_val(var)))
+            .collect();
+
+        let status = self.problem.get_status();
+        match status {
+            russcip::status::Status::OPTIMAL => Ok(SCIPSolution { values }),
+            russcip::status::Status::INFEASIBLE => {
+                return Err(ResolutionError::Infeasible);
+            }
+            russcip::status::Status::UNBOUNDED => {
+                return Err(ResolutionError::Unbounded);
+            }
+            other_status => {
+                return Err(ResolutionError::Str(format!(
+                    "Unexpected status {:?}",
+                    other_status
+                )));
+            }
+        }
+    }
+
+    fn add_constraint(&mut self, c: Constraint) -> ConstraintReference {
+        let constant = -c.expression.constant;
+        let lhs = match c.is_equality {
+            true => constant,
+            false => -f64::INFINITY,
+        };
+
+        let mut vars = HashMap::new();
+        for var in self.problem.get_vars() {
+            let var_index = var.get_index();
+            vars.insert(var_index, var);
+        }
```

Comment:
You are building and destroying a hashmap containing all the problem variables each time a constraint is added. This is not going to scale well when there are many variables and contraints. You should build this mapping once and for all if it is needed. You can look at how it is done in the minilp binding, for instance.

Reply:
Thanks! I changed it to building the mapping once in the beginning.